app status test assertions

joe-kimmel-vmw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

If this is your first time, please read our contributor guidelines: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/CONTRIBUTING.md and developer guide https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/docs/dev.md
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
I think this test just happens to run first (`go test` has an undefined order, technically) but because the first thing that happens is a `kapp deploy` with an allowError=true, it means that you can get some funny errors if there's stale testing apps deployed in your test namespace. This is especially obscured if you happen to be making a change to the App Status object because then you'll really be wondering what you broke.  So these assertions will hopefully help someone recognize the root cause of the errors they're seeing faster than I did today.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. 

-->
```release-note

```

#### Additional Notes for your reviewer:

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

joe-kimmel-vmw committed:
app status test assertions

cppforlife commented on test/e2e/kappcontroller/app_status_test.go:
```diff
@@ -58,10 +59,17 @@ spec:
 	logger.Section("deploy", func() {
-		kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name}, e2e.RunOpts{StdinReader: strings.NewReader(appYaml), AllowError: true})
+		out, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name}, e2e.RunOpts{StdinReader: strings.NewReader(appYaml), AllowError: true})
+		// it's supposed to error, but it's also supposed to create the three supporting resources successfully:
+		assert.Error(t, err)
+		assert.Contains(t, out, "ok: reconcile role/kappctrl-e2e-ns-role (rbac.authorization.k8s.io/v1) namespace: kappctrl-test")
+		assert.Contains(t, out, "ok: reconcile rolebinding/kappctrl-e2e-ns-role-binding (rbac.authorization.k8s.io/v1) namespace: kappctrl-test")
+		assert.Contains(t, out, "ok: reconcile serviceaccount/kappctrl-e2e-ns-sa (v1) namespace: kappctrl-test")
+
```
namespace cant be hard coded since it comes from kapp class. given below assertion im not sure it's worth adding this. alternatively you can may be assert on error message from below inside expectedStatus.

joe-kimmel-vmw replied:
i'm just doing substring assertions so it's appealing to amend by shortening the string to exclude the namespace.   I'd still like to capture the information that those three resources should be successfully created either by asserting here or below in the output of the inspect.

cppforlife replied:
not pretty looking, but ok.

joe-kimmel-vmw committed:
Update app_status_test.go

shorten substring assertions to exclude namespace and version info so we're just asserting that something with that kind/name reconciled successfully

cppforlife commented on test/e2e/kappcontroller/app_status_test.go:
```diff
@@ -58,10 +59,17 @@ spec:
+		assert.Contains(t, out, "ok: reconcile role/kappctrl-e2e-ns-role")
+		assert.Contains(t, out, "ok: reconcile rolebinding/kappctrl-e2e-ns-role-binding")
+		assert.Contains(t, out, "ok: reconcile serviceaccount/kappctrl-e2e-ns-sa")
+
```
nit: newline

joe-kimmel-vmw committed:
Update app_status_test.go

remove superfluous newline
